server/filestore.go

```diff
@@ -3791,6 +3869,166 @@ func (fs *fileStore) writeMsgRecord(seq uint64, ts int64, subj string, hdr, msg
 	return rl, err
 }
 
+func (mb *msgBlock) recompressOnDiskIfNeeded() error {
+	// Wait for disk I/O slots to become available. This prevents us from
+	// running away with system resources.
+	<-dios
+	defer func() {
+		dios <- struct{}{}
+	}()
+
+	alg := mb.fs.fcfg.Compression
+	mb.mu.Lock()
+	defer mb.mu.Unlock()
+
+	origFN := mb.mfn         // The original message block on disk.
+	oldFN := mb.mfn + ".old" // The original message block will be moved here.
+	tmpFN := mb.mfn + ".tmp" // The compressed block will be written here.
+
+	// Open up the file block and read in the entire contents into memory.
+	// One of two things will happen:
+	// 1. The block will be compressed already and have a valid metadata
+	//    header, in which case we do nothing.
+	// 2. The block will be uncompressed, in which case we will compress it
+	//    and then write it back out to disk, reencrypting if necessary.
+	origFD, err := os.OpenFile(origFN, os.O_RDWR, defaultFilePerms)
+	if err != nil {
+		return err
+	}
+	origBuf, err := io.ReadAll(origFD)
+	if err != nil {
+		return fmt.Errorf("failed to read original block from disk: %w", err)
+	}
+	if err := origFD.Close(); err != nil {
+		return fmt.Errorf("failed to close original file: %w", err)
+	}
+
+	// If the block is encrypted then we will need to decrypt it before
+	// doing anything. We always encrypt after compressing because then the
+	// compression can be as efficient as possible on the raw data, whereas
+	// the encrypted ciphertext will not compress anywhere near as well.
+	// The block encryption also covers the optional compression metadata.
+	if mb.bek != nil && len(origBuf) > 0 {
+		bek, err := genBlockEncryptionKey(mb.fs.fcfg.Cipher, mb.seed, mb.nonce)
+		if err != nil {
+			return err
+		}
+		mb.bek = bek
+		mb.bek.XORKeyStream(origBuf, origBuf)
+	}
+
+	meta := &CompressionInfo{}
+	if _, err := meta.UnmarshalMetadata(origBuf); err != nil {
+		// An error is only returned here if there's a problem with parsing
+		// the metadata. If the file has no metadata at all, no error is
+		// returned and the algorithm defaults to no compression.
+		return fmt.Errorf("failed to read existing metadata header: %w", err)
+	}
+	if meta.Algorithm == alg {
```

Comment:
If we think this is common should we just peek at this part before loading in the whole block?